presto-main/src/main/java/io/prestosql/sql/planner/LocalExecutionPlanner.java

```diff
@@ -2235,6 +2241,26 @@ public PhysicalOperation visitSemiJoin(SemiJoinNode node, LocalExecutionPlanCont
             int probeChannel = probeSource.getLayout().get(node.getSourceJoinSymbol());
             int buildChannel = buildSource.getLayout().get(node.getFilteringSourceJoinSymbol());
 
+            ImmutableList.Builder<OperatorFactory> buildOperatorFactories = new ImmutableList.Builder<>();
+            buildOperatorFactories.addAll(buildSource.getOperatorFactories());
+
+            node.getDynamicFilterId().ifPresent(filterId -> {
+                // Add a DynamicFilterSourceOperatorFactory to build operator factories
+                log.debug("[Semi-join] Dynamic filter: %s", node.getDynamicFilterId().get());
+                LocalDynamicFilterConsumer filterConsumer = new LocalDynamicFilterConsumer(
+                        ImmutableMap.of(filterId, buildChannel),
+                        ImmutableMap.of(filterId, buildSource.getTypes().get(buildChannel)),
+                        1);
+                addSuccessCallback(filterConsumer.getDynamicFilterDomains(), context::addDynamicFilter);
+                buildOperatorFactories.add(new DynamicFilterSourceOperatorFactory(
+                        buildContext.getNextOperatorId(),
+                        node.getId(),
+                        filterConsumer.getTupleDomainConsumer(),
+                        ImmutableList.of(new DynamicFilterSourceOperator.Channel(filterId, buildSource.getTypes().get(buildChannel), buildChannel)),
+                        getDynamicFilteringMaxPerDriverRowCount(context.getSession()),
```

Comment:
Since semi-join build side is not partitioned right now, I think it will be better to use higher limits on filter size here (maybe 4x).

Reply:
Makes sense. This could be separate PR. Maybe we want to rethink per driver limits and use node global limit (which we could divide between drivers).